pulp_container/app/registry_api.py

```diff
@@ -656,8 +656,9 @@ def put(self, request, path, pk=None):
                 )
                 blob_artifact.save()
             except IntegrityError:
-                pass
-
+                ca = ContentArtifact.objects.get(content=blob, relative_path=digest)
+                ca.artifact = artifact
+                ca.save(update_fields=["artifact"])
```

Comment:
Trying to understand this...
The integrity error will be called on the condition that `(content, relative_path)` must be unique. So we trigger this exception handler when such a content artifact exists already. Now we look for it and it may (immediate) or may not (on demand, reclaimed) already point to that artifact. Does the call to `save` check whether a SQL call to the database is necessary?

What if after receiving the `IntegrityError`, the `ContentArtifact` is deleted? Can we do a `select_for_update` first? [maybe a separate refactoring effort]

Reply:
what is the real case scenario when content artifact is deleted?

Reply:
> Trying to understand this... The integrity error will be called on the condition that `(content, relative_path)` must be unique. So we trigger this exception handler when such a content artifact exists already. Now we look for it and it may (immediate) or may not (on demand, reclaimed) already point to that artifact. Does the call to `save` check whether a SQL call to the database is necessary?
i don't think there is any check, so i think we'd have one unnecessary update. see https://github.com/pulp/pulp_container/pull/596#issuecomment-1047762501
> 
> What if after receiving the `IntegrityError`, the `ContentArtifact` is deleted? Can we do a `select_for_update` first? [maybe a separate refactoring effort]

Reply:
> what is the real case scenario when content artifact is deleted?

I was thinking about orphan cleanup. But that should not be a problem anymore once we arrive at this line.

Reply:
orphan cleanup would not be an issue, artifact and blob has been just touched.